src/kbc_modules/cc_kbc/crypto/mod.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2022 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use aes_gcm::aead::{Aead, NewAead};
+use aes_gcm::Aes256Gcm;
+use anyhow::*;
+use rsa::pkcs8::EncodePublicKey;
+use rsa::{PaddingScheme, RsaPrivateKey, RsaPublicKey};
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha384};
+
+const RSA_ALGORITHM: &str = "rsa-pkcs1v15";
+const RSA_PUBKEY_LENGTH: usize = 2048;
+const NEW_PADDING: fn() -> PaddingScheme = PaddingScheme::new_pkcs1v15_encrypt;
+
+const AES_GCM_256_ALGORITHM: &str = "aes-gcm-256";
+
+// The key inside TEE to decrypt confidential data.
+#[derive(Debug, Clone)]
+pub struct TeeKey {
+    private_key: RsaPrivateKey,
+    public_key: RsaPublicKey,
+}
+
+// The public key of TEE.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct TeePubKey {
+    algorithm: String,
+    #[serde(rename = "pubkey-length")]
+    pubkey_length: usize,
+
+    // public key string in PEM format.
+    pubkey: String,
+}
+
+impl TeeKey {
+    pub fn new() -> Result<TeeKey> {
+        let mut rng = rand::thread_rng();
+
+        let private_key = RsaPrivateKey::new(&mut rng, RSA_PUBKEY_LENGTH)?;
+        let public_key = RsaPublicKey::from(&private_key);
+
+        Ok(TeeKey {
+            private_key,
+            public_key,
+        })
+    }
+
+    // Export TEE public key as specific structure.
+    pub fn export_pubkey(&self) -> Result<TeePubKey> {
+        let pem_line_ending = rsa::pkcs1::LineEnding::default();
+        let pubkey_pem_string = self.public_key.to_public_key_pem(pem_line_ending)?;
+
+        Ok(TeePubKey {
+            algorithm: RSA_ALGORITHM.to_string(),
+            pubkey_length: RSA_PUBKEY_LENGTH,
+            pubkey: pubkey_pem_string,
+        })
+    }
+
+    // Use TEE private key to decrypt cipher text.
+    pub fn decrypt(&self, cipher_text: Vec<u8>) -> Result<Vec<u8>> {
+        let padding = NEW_PADDING();
+
+        self.private_key
+            .decrypt(padding, &cipher_text)
+            .map_err(|e| anyhow!("TEE rsa key decrypt failed: {:?}", e))
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct CryptoAnnotation {
+    algorithm: String,
+
+    // IV required by some symmetric encryption algorithms.
+    // Base64 encoded.
+    // If don't need, left blank.
+    iv: String,
+
+    // The symmetric key encrypted by TEE's public key.
+    // This symmetric key is used to encrypt API output by KBS.
+    // Base64 encoded.
+    #[serde(rename = "enc-symkey")]
+    enc_symkey: String,
+}
+
+impl CryptoAnnotation {
+    // Use the TEE private key to unwrap the encrypted symmetric key,
+    // then use the symmetric key to decrypt cipher text.
```

Comment:
Is the symmetric key going to be randomized for each response? Otherwise it seems like we could just decrypt it once and store it in the KBC alongside the token.

Reply:
Whether the symmetric key is randomized for each response depends on the security policy adopted by KBS, but I think maybe it is not a good idea to save this key in KBC, because it will cause us to design additional communication fields to update this key (for example, when KBS finds that there is a risk of key disclosure).

Reply:
Ok. Makes sense to leave it up to the KBS. Is there anything that prevents replay of a response?

Reply:
The replay attack defense of response is the responsibility of the protocol that transmits these messages (HTTPS). These JSON structures defined by us will be passed as HTTPS payload.

Reply:
In order to avoid maliciously impersonating KBS, KBS's TLS certificate need to be preset in the environment in which KBC runs, but I'm not sure how to achieve this currently...

Reply:
Yeah we have a bit of a chicken-and-egg problem here. It seems like we want the KBS identity to be measured somehow, but since the KBC is the one that validates the measurement this might not actually give us the guarantee that we want.

One potential answer is that we need to securely expose the KBS identity to the guest owner directly or to other services operating on behalf of the guest owner. In other words, we need to extend the chain of trust from the HW to the KBS to some other entity like a database providing secure data.

This is kind of a complicated problem that we probably don't need to worry about here yet. I am planning to make another issue about it soon.